File: app/src/main/java/me/kirimin/mitsumine/util/EntryInfoFunc.java
package me.kirimin.mitsumine.util;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

import me.kirimin.mitsumine.model.Bookmark;
import me.kirimin.mitsumine.model.EntryInfo;
import rx.functions.Func1;

public class EntryInfoFunc {

    public static Func1<JSONObject, EntryInfo> mapToEntryInfo() {
        return new Func1<JSONObject, EntryInfo>() {
            @Override
            public EntryInfo call(JSONObject response) {
                try {
                    String title = response.getString("title");
                    int count = response.getInt("count");
                    String url = response.getString("url");
                    String thumbnail = response.getString("screenshot");
                    List<Bookmark> bookmarkList = parseBookmarkList(response.getJSONArray("bookmarks"));
                    return new EntryInfo(title, count, url, thumbnail, bookmarkList);
                } catch (JSONException e) {
                    return null;
                }
            }
        };
    }

    public static Func1<EntryInfo, Boolean> isNotNullEntryInfo() {
        return new Func1<EntryInfo, Boolean>() {
            @Override
            public Boolean call(EntryInfo entryInfo) {
                return entryInfo != null;
            }
        };
    }

    public static Func1<Bookmark, Boolean> hasComment() {
        return new Func1<Bookmark, Boolean>() {
            @Override
            public Boolean call(Bookmark bookmark) {
                return !bookmark.getComment().equals("");
            }
        };
    }

    public static Func1<JSONObject, Bookmark> mapToMyBookmarkInfo() {
        return new Func1<JSONObject, Bookmark>() {
            @Override
            public Bookmark call(JSONObject jsonObject) {
                try {
                    String user = jsonObject.getString("user");
                    String comment = jsonObject.getString("comment");
                    String timeStamp = jsonObject.getString("created_datetime");
                    JSONArray tagJsonArray = jsonObject.getJSONArray("tags");
                    List<String> tags = new ArrayList<>();
                    for (int i = 0; i < tagJsonArray.length(); i++) {
                        tags.add(tagJsonArray.getString(i));
                    }
                    return new Bookmark(user, tags, timeStamp, comment, "");
                } catch (JSONException e) {
                    return null;
                }
            }
        };
    }

    private static List<Bookmark> parseBookmarkList(JSONArray bookmarks) throws JSONException {
        List<Bookmark> bookmarkList = new ArrayList<>();
        for (int i = 0; i < bookmarks.length(); i++) {
            JSONObject bookmark = bookmarks.getJSONObject(i);
            String user = bookmark.getString("user");
            String comment = bookmark.getString("comment");
            String timeStamp = bookmark.getString("timestamp");
            timeStamp = timeStamp.substring(0, timeStamp.indexOf(" "));
            String userIcon = "http://n.hatena.com/" + user + "/profile/image.gif?type=face&size=64";
            List<String> tags = parseTags(bookmark.getJSONArray("tags"));
            bookmarkList.add(new Bookmark(user, tags, timeStamp, comment, userIcon));
        }
        return bookmarkList;
    }

    private static List<String> parseTags(JSONArray tags) throws JSONException {
        List<String> results = new ArrayList<>();
        for (int i = 0; i < tags.length(); i++) {
            results.add(tags.getString(i));
        }
        return results;
    }
}

File: app/src/main/java/me/kirimin/mitsumine/util/EntryInfoFunc.kt
package me.kirimin.mitsumine.util

import android.text.Html

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

import java.util.ArrayList
import java.util.regex.Matcher
import java.util.regex.Pattern

import me.kirimin.mitsumine.model.Bookmark
import me.kirimin.mitsumine.model.EntryInfo
import rx.functions.Func1
import toList

public class EntryInfoFunc {
    companion object {

        private val urlLinkPattern = Pattern.compile("(http://|https://){1}[\\w\\.\\-/:\\#\\?\\=\\&\\;\\%\\~\\+]+", Pattern.CASE_INSENSITIVE)

        public fun toEntryInfo(response: JSONObject): EntryInfo {
            try {
                val title = response.getString("title")
                val count = response.getInt("count")
                val url = response.getString("url")
                val thumbnail = response.getString("screenshot")
                val bookmarkList = parseBookmarkList(response.getJSONArray("bookmarks"))
                return EntryInfo(title, count, url, thumbnail, bookmarkList)
            } catch (e: JSONException) {
                return EntryInfo()
            }
        }

        public fun hasComment(bookmark: Bookmark): Boolean {
            return !bookmark.comment.toString().isEmpty()
        }

        public fun mapToMyBookmarkInfo(): Func1<JSONObject, Bookmark> {
            return object : Func1<JSONObject, Bookmark> {
                override fun call(jsonObject: JSONObject): Bookmark? {
                    try {
                        val user = jsonObject.getString("user")
                        val comment = jsonObject.getString("comment")
                        val timeStamp = jsonObject.getString("created_datetime")
                        val tags = jsonObject.getJSONArray("tags").toList<String>()
                        val bookmark = Bookmark(user, tags, timeStamp, comment, "")
                        bookmark.setPrivate(jsonObject.getBoolean("private"))
                        return bookmark
                    } catch (e: JSONException) {
                        return null
                    }

                }
            }
        }

        throws(JSONException::class)
        private fun parseBookmarkList(bookmarks: JSONArray): List<Bookmark> {
            return bookmarks.toList<JSONObject>().map { bookmark ->
                val user = bookmark.getString("user")
                val comment = parseCommentToHtmlTag(bookmark.getString("comment"))
                val timeStampTmp = bookmark.getString("timestamp")
                val timeStamp = timeStampTmp.substring(0, timeStampTmp.indexOf(" "))
                val userIcon = "http://cdn1.www.st-hatena.com/users/" + user.subSequence(0, 2) + "/" + user + "/profile.gif"
                val tags = bookmark.getJSONArray("tags").toList<String>()
                Bookmark(user, tags, timeStamp, comment, userIcon)
            }
        }

        private fun parseCommentToHtmlTag(comment: String): CharSequence {
            val matcher = urlLinkPattern.matcher(comment)
            return Html.fromHtml(matcher.replaceAll("<a href=\"$0\">$0</a>"))
        }
    }
}
